src/middlewares/__tests__/destTransformCompactedPayloadV1.test.ts

```
@@ -0,0 +1,112 @@
+    },
+    libraries: [],
+    credentials: [],
+  };
+  it('should not alter body if http header is not present', async () => {
+    const originalBody = [
+      {
+        ...input,
+        destination: destination,
+        connection: connection,
+      },
+    ] as ProcessorTransformationRequest[];
+    const ctx = {
+      request: {
+        body: originalBody,
+      },
+    } as Koa.Context;
+    ctx.get = jest.fn((v: string) => '');
+
+    DestTransformCompactedPayloadV1Middleware.handle(ctx, async () => {
+      expect(ctx.request.body).toEqual(originalBody);
+    });
+  });
+
+  it('should alter body if http header is present', async () => {
+    const expectedBody = [
+      {
+        ...input,
+        destination: destination,
+        connection: connection,
+      },
+    ] as ProcessorTransformationRequest[];
+
+    const ctx = {
+      request: {
+        body: {
+          input: [
+            {
+              message: { type: 'track' },
+              metadata: {
+                sourceId: 'sourceId',
+                destinationId: 'destinationId',
+                messageId: 'messageId',
+                workspaceId: 'workspaceId',
+                sourceType: 'sourceType',
+                sourceCategory: 'sourceCategory',
+                jobId: 1,
+                destinationType: 'destinationType',
+              },
+              libraries: [],
+              credentials: [],
+            },
+          ],
+          destinations: {
+            destinationId: destination,
+          },
+          connections: {
+            'sourceId:destinationId': connection,
+          },
+        } as ProcessorCompactedTransformationRequest,
+      },
+    } as Koa.Context;
+    ctx.get = jest.fn((v) => (v === 'x-content-format' ? 'json+compactedv1' : ''));
+
+    DestTransformCompactedPayloadV1Middleware.handle(ctx, async () => {
+      expect(ctx.request.body).toEqual(expectedBody);
+    });
+  });
+});
```

We can add test cases for mismatch of destination or connection config id.
Ideal case is to throw an error with clear logs.

---

Are there any validations now on the normal flow for `connection` or `destination` objects, e.g. that they are properly filled?

---

imo, these would be most beneficial, because they would handle both scenarios, of compacted and non-compacted payloads.

---

Assuming that a combination is missing, the result would be an empty destination or connection field.
My point is, do we protect ourselves now against empty destination or connection fields? If not, why should we do it here at this middleware?

ps. integration with rudder server will be verified through rudder-server's integration tests with rudder-transformer, i.e. this feature will be enabled while running all rudder-server tests

---

If validations aren't in place, any observability measures can we implement to ensure that the lookup never returns empty?

For example, in the case of Braze, if the destination configuration is missing the [dataCenter](https://github.com/rudderlabs/rudder-transformer/blob/35cda1ed1b49eee6548a55d04d0711dff83055a6/src/v0/destinations/braze/transform.js#L210) field, instrumentation errors will be thrown. If the lookup returns an empty destConfig, the same errors will be thrown. How can we differentiate between customer misinstrumentation and a potential bug in the middleware?

---

Rudder server's code preparing these payloads will be quite straightforward and we are going to test all affected endpoints through rudder-server tests.
IMO validation at the middleware layer is not needed, but if you believe that validation about keys missing is required, I shall add it

---

Are we sure that we are not going to introduce some regression though? E.g. do ALL transformations require a destination/connection object and WILL FAIL if it is missing?
I still believe that these kind of validations at this layer are not a good idea and could cause more issues than they would try to solve.

---

If that happens, this is an edge case and platform bug, and we should throw a 500 error from the transformer. In most cases we easily cover this using unit tests and some integration tests. we don't need to put validations, anyways if that happens it will fail while processing so we don't need to explicitly put validations during event processing as it is costly

---

@atzoum as we start this new process now, initially to be on the safe side let's go with validations and throw specific errors (PlatformError) so that we don't accidentally end up aborting events when it is a platform issue. we can't control 200+ integrations' behavior, they might behave differently as over time different people developed them and there is no consistency in them. Once we are confident about this we can remove validations for efficiency.

---

I added some validation with respect to connections and destinations as requested.
`@koladilip` I am using a `PlatformError` and explicitly setting the error code to `500` because my guess was that the default one is `400`. Am I missing something important? Are we using some special middleware which performs special handling for PlatformErrors? What is the error code we should return here? Codebase seems a bit confusing to me...

---

PlatformError with `500` should be right one here so that events get retried. But to know this behaviour, [platform-errors](https://github.com/rudderlabs/rudderstack-operator/blob/master/operator-helm/alerts/cross-customer-pipelines/integrations/platform-errors.yaml) alert should be triggered - @koladilip should the statTags be explicitly added here as the alert definition depends on `errorCategory` key ?

---

https://github.com/rudderlabs/rudderstack-operator/blob/f98fbb703c3522c397a9bd18c5b3e0f95a1a6245/operator-helm/alerts/cross-customer-pipelines/integrations/platform-errors.yaml#L3

https://github.com/rudderlabs/rudder-integrations-lib/blob/fe534c8349f7e85d7a27dd2ca8e0369987e7bf61/src/errors/platform_error.ts#L4

errorCategory=platform is automatically added